CI/CD 구축(#16)

chanwoothecali opened the pull request:
## 젠킨스를 이용한 CI/CD 구축
- ncloud를 이용한 서버 구축
- jenkins multibranch pipeline과 jenkinsfile을 이용한 CI/CD 자동화

chanwoothecali committed:
jenkin test

chanwoothecali committed:
jenkins test 실행어 변경

chanwoothecali committed:
jenkinsfile env 수정

chanwoothecali committed:
jenkinsfile 수정

ChaYujin committed:
local,product,test 프로파일 분리, mysql dependency 추가

- 각 환경별로 다른 DB접근정보가 필요하여 local, product, test로 프로파일 분리함
  * local : local에 설치된 H2서버에 연결
  * product : url, username, password를 외부에서 받고 mysql에 연결하도록 설정
  * test : 인메모리 H2서버에 연결
- mysql사용시 필요한 의존성 정보 추가

ChaYujin committed:
Jenkinsfile의 빌드 스테이지 명령어 수정

./gradlew clean build -x test로 test단계 없이 빌드되도록 명령어 수정

ChaYujin committed:
Jenkinsfile을 Scripted Pipeline로 수정

간단한 들여쓰기와 확장된 기능 사용을 위해 Declarative Pipeline->Scropted Pipeline으로 변경

ChaYujin committed:
Jenkinsfile의 stage directive안의 steps제거

scripted pipeline은 steps directive를 사용하지 않으므로 제거함

ChaYujin committed:
Jenkinsfile에 Deploy stage추가

Deploy stage 추가
 - ssh agent툴로 ssh연결
 - deploy_server정보로 ssh연결

* 참고로 pwd정보에 따라 업로드할 jar위치 변경 필요

ChaYujin committed:
Jenkinsfile 수정 - 정의한 변수로 값이 설정되도록 수정

정의한 변수로 값이 설정되도록 홑따옴표를 사용하도록 수정

ChaYujin committed:
Jenkinsfile 수정 - sshagent안의 sh명령어 호출시 쌍따옴표를 쓰도록 수정

- 변수값이 세팅되어 명령어가 실행되도록 홑따옴표가 아닌 쌍따옴표로 실행 명령어를 지정함
- jar파일 업로드, deploy.sh파일 실행하는 명령어도 수행하도록 주석 제거

ChaYujin committed:
프로파일 - 프로파일 기본값 세팅, DB정보가 필요한 TEST코드에 @ActiveProfiles추가

- 프로파일 옵션이 들어오지 않은 경우 동작할 프로파일 기본값을 local로 설정함
- DB접근이 필요한 즉, test프로파일의 DB정보가 필요한 테스트 코드에 @ActiveProfiles("test")를 추가함

chanwoothecali committed:
jenkinsfile 환경변수 설정

jenkinsfile이 jenkins 내에 등록된 환경 변수를 참조하도록 수정

chanwoothecali committed:
jenkinsfile CD 부분 수정

develop 브랜치에서만 CD가 일어나도록 수정

ChaYujin committed:
Jenkinsfile - Parameter Setting 스텝 추가, 배포서버 실행스크립트에 paroduct 파라메터 추가

젠킨스 환경변수로 지정해놓은 DB 정보를 세팅하는 스텝을 추가하고, 배포서버에 서버 띄우는 스크립스 실행시 profile 정보를 넘겨주도록 수정

ChaYujin committed:
jenkinsfile - ${} 특수문자가 그대로 인지되도록 수정

${} 특수문자가 인지되도록 '\' 추가

ChaYujin committed:
jenkinsfile - 백슬래시 특수문자 제거

${} 인지를 위해 넣었던 백슬래시로 인해 시멘틱에러가 나는 문제로 제거처리함

ChaYujin committed:
jenkinsfile - DB값 치환 명령어 수정

$앞쪽과  }특수문자 앞에 백슬래시 추가, sed 2번째 파라메터인 파일명에는 쌍따옴표 제거

ChaYujin committed:
Jenkinsfiles - 오류로 인해 '}'앞 백슬래시 제거

스크립트 실행 오류로 '}'앞의 백슬래시 제거

ChaYujin committed:
Jenkinsfiles - 특수문자 인식을 위해 '/'대신 '%'로 변경

특수문자 인식을 위해 '/'대신 '%'로 변경

ChaYujin committed:
Jenkinsfile - 쌍따옴표를 홑따옴표로 변경

홑따옴표로 문자열 그대로 들어가도록 수정해봄

ChaYujin committed:
Jenkinsfile - 문자열치환대신 deploy.sh 파일 변수로 넣어주도록 수정

deploy.sh에서 -D옵션으로 스프링 환경변수정보를 넣어줄 수 있다. deploy.sh에서 넘겨받은 변수를 각각 -DDB_URL, -DDB_USERNAME, -DDB_PASSWOTD로 스프링에 환경변수를 넘겨주도록 수정하였다.

ChaYujin committed:
Jenkinsfile - deveop브랜치에서만 배포작업이 일어나도록 주석 제거

배포 테스트를 위해 주석처리해놓았던 when 조건문을 주석해제함

ChaYujin committed:
Jenkinsfile - CICD 성공/실패 여부 PR반영하는 코드 추가, Deploy단계 동작 조건 변경

- 젠킨스에 github hook HTTP REQUEST로 넘어오는 json 정보를 파싱해서 PR_STATUS와 MERGED로 정의해둠
- CI/CD 단계 처리시 성공/실패 여부를 처리할 setBuildStatus()함수를 추가하고, 이를 호출하는 post{} 에 동작 조건을 PR이 open됐을 때로 지정함
- Deploy단계의 동작 조건을 PR이 closed되면서 머지가 되는 시점에 Deploy하도록 지정함

* 참고사이트 : https://bit-basics-backup.netlify.app/2019/september/jenkins-job-github-pullrequest/
* 참고사이트 : https://docs.github.com/ko/webhooks-and-events/webhooks/webhook-events-and-payloads?actionType=opened#pull_request

chanwoothecali committed:
jenkinsfile 조건문 변경

chanwoothecali committed:
jenkinsfile when 구문 삭제

f-lab-michael commented on src/main/resources/application.yml:
```diff
@@ -1,10 +1,45 @@
+spring:
+  config:
+    activate:
+      on-profile: "local"
```
application-local.yml로 선언하면 같은 효과를 가질수있습니다.

chanwoothecali replied:
넵 분리하겠습니다!

f-lab-michael commented on src/main/resources/application.yml:
```diff
@@ -1,10 +1,45 @@
+spring:
+  config:
+    activate:
+      on-profile: "product"
```
application-product.yml 로 선언하면 같은효과를 가질수있습니다.

f-lab-michael commented on src/main/resources/application.yml:
```diff
@@ -1,10 +1,45 @@
+spring:
+  config:
+    activate:
+      on-profile: "test"
```
얘도 똑같음

f-lab-michael commented on src/test/java/com/flab/modu/market/repository/MarketRepositoryTest.java:
```diff
@@ -12,8 +12,10 @@
+import org.springframework.test.context.ActiveProfiles;
 
 @DisplayName("Market JPA 연결 테스트")
+@ActiveProfiles("test")
```
모든 테스트 파일에 프로파일을 엑티브 하는것보다 기본설정에 h2디비 설정을 하고 local, prod 에서 설정을 덮어 쓰는 형태로 하는게 어떨까요?

chanwoothecali replied:
해당 애노테이션 제거하고 Jenkinsfile에 SPRING_PROFILES_ACTIVE=test 옵션을 설정해 test profile을 따라가게 만들었습니다.

chanwoothecali committed:
profile 분리 및 ActiveProfile 제거

- application.yml profile 별로 분리
- ActiveProfile 애노테이션 제거

chanwoothecali committed:
jenkinsfile test 명령어 수정

test시 product 프로파일을 보도록 수정

chanwoothecali committed:
gradle test시 환경변수 추가

chanwoothecali committed:
jenkinsfile test log

추후 롤백예정

chanwoothecali committed:
test push

추후 롤백 예정

chanwoothecali committed:
jenkinsfile 테스트용 변수출력

chanwoothecali committed:
jenkins test profile 설정

- jenkinsfile test시 test profile을 따라가도록 설정
- 기존 환경변수 삭제

chanwoothecali committed:
jenkinsfile test profile 명령어 변경

chanwoothecali committed:
local profile logging level 변경

local 테스트 시 log 추가

ChaYujin commented on src/main/resources/application-local.yml:
```diff
@@ -0,0 +1,16 @@
+spring:
+  datasource:
+    url: jdbc:h2:mem:modu
```
datasource의 url의 내용이 기존의 local 설정과 다르게 되어있습니다.
tcp localhost 연결에서 인메모리 h2로 의도적으로 변경하신걸까요~?

chanwoothecali replied:
해당 내용 테스트만 하고 원복한다는 게 그렇게 됐군요. 제가 현재 회사라 수정 부탁드리겠습니다!

ChaYujin committed:
test패키지 application.yml추가

test패키지 내에서 사용할 application.yml을 추가함. 
이로 인해 gradle test 를 진행할 때, 별도의 환경변수는 필요하지 않아 Jenkinsfile의 Test단계의 환경변수를 제거함

ChaYujin committed:
local 프로파일의 DB URL정보를 원래대로 변경

application.yml 파일를 분리하는 과정에서 local프로파일의 DB정보가 변경되어 원래대로 수정함

ChaYujin committed:
파일 마지막 줄에 뉴라인 추가

EOL 에러가 나지 않도록 마지막 줄에 뉴라인 추가함

ChaYujin committed:
디폴트 프로파일 정보를 local로 변경, 배포환경 프로파일명을 원래대로 product로 변경

- test환경에서 사용하는 applcation.yml파일을 test패키지로 이동시켰기 때문에 test환경에서 사용하는 정보는 불필요하므로 기본 프로파일을 local로 변경함
- local환경에서 테스트시 테이블을 드랍시키고 다시 생성할 필요가 없으므로 jpa.hibernate.ddl-auto값을 none으로 수정함
- yml파일이 분리되면서 배포환경 프로파일명이 product에서 prod로 변경되었었다. 하지만 Jenkinsfile의 배포단계에서 product로 사용하고 있으므로 다시 product로 맞춰 변경함

ChaYujin reviewed (approved):
모든 코드 확인했습니다. 👍
